src/container/srv_container.c

```diff
@@ -1893,34 +1905,38 @@ cont_close_hdls(struct cont_svc *svc, struct cont_tgt_close_rec *recs,
 	if (rc != 0)
 		D_GOTO(out, rc);
 
-	/*
-	 * Use one TX per handle to avoid calling ds_cont_epoch_fini_hdl() more
-	 * than once in a TX, in which case we would be attempting to query
-	 * uncommitted updates. This could be optimized by adding container
-	 * UUIDs into recs[i] and sorting recs[] by container UUIDs. Then we
-	 * could maintain a list of deleted LREs and a list of deleted LHEs for
-	 * each container while looping, and use the lists to update the GHCE
-	 * once for each container. This approach enables us to commit only
-	 * once (or when a TX becomes too big).
-	 */
-	for (i = 0; i < nrecs; i++) {
-		struct rdb_tx tx;
+	rc = rdb_tx_begin(svc->cs_rsvc->s_db, svc->cs_rsvc->s_term, &tx);
+	if (rc != 0)
+		goto out;
 
-		rc = rdb_tx_begin(svc->cs_rsvc->s_db, svc->cs_rsvc->s_term,
-				  &tx);
-		if (rc != 0)
-			break;
+	for (i = 0; i < nrecs; i++) {
 		rc = cont_close_one_hdl(&tx, svc, ctx, recs[i].tcr_hdl);
-		if (rc != 0) {
+		if (rc != 0)
+			goto out_tx;
+
+		/*
+		 * Yield frequently, in order to cope with the slow
+		 * vos_obj_punch operations invoked by rdb_tx_commit for
+		 * deleting the handles. (If there is no other RDB replica, the
+		 * TX operations will not yield, and this loop would occupy the
+		 * xstream for too long.)
+		 */
+		if ((i + 1) % 32 == 0) {
+			rc = rdb_tx_commit(&tx);
```

Comment:
only a question so I can better understand the implementation. If rdb_tx_commit() fails (and let's say i=31 for discussion) then the IV invalidate for 32 handles will have been sent. So while this transaction at the pool/container service leader is going to fail, part of the handle closing work has occurred is that right?

I guess in the original code the partial work is limited to just one handle impacted since there is no batching?

Reply:
@kccain,

> only a question so I can better understand the implementation. If rdb_tx_commit() fails (and let's say i=31 for discussion) then the IV invalidate for 32 handles will have been sent. So while this transaction at the pool/container service leader is going to fail, part of the handle closing work has occurred is that right?

That's right. I wondered about this scenario too. No clear idea how to address it at the moment.

> I guess in the original code the partial work is limited to just one handle impacted since there is no batching?

Hmm, the original code would invalidate all N handles, then actually delete just the first one in the worst case.